Parameterize ConstantVectorSource

avalenzu opened the pull request:
This PR replaces the `source_value_` member variable in `ConstantVectorSource` with a numeric parameter. This allows one diagram to represent a family of models that differ only in the value of the source.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/6660)
<!-- Reviewable:end -->

avalenzu commented:
+@jwnimmer-tri for feature review

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6660#-:-KpuZyZvsMoFJ5-qU_Fv:b-982cs4)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Sorry for the delay, my reviewable dashboard stopped updating and I didn't notice.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

+@EricCousineau-TRI for platform review, please.

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/primitives/constant_vector_source.cc, line 14 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6660#-Kpv1ThhBGBtMMdToxbV:-Kpv1ThiAX10AQPyEvkQ:b-2yk851) ([raw file](https://github.com/robotlocomotion/drake/blob/ab8c0949b0aef74a55f9acaf6006796be443aaaa/drake/systems/primitives/constant_vector_source.cc#L14)):*
> ```cpp
> ConstantVectorSource<T>::ConstantVectorSource(
>     const Eigen::Ref<const VectorX<T>>& source_value)
>     : ConstantVectorSource(BasicVector<T>(source_value)) {}
> ```

FYI Nice!  (Everyone should strongly prefer delegating constructors, instead of independently-implemented ones.)

---

*[drake/systems/primitives/constant_vector_source.cc, line 33 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6660#-Kpv1km_FLVWgTIdO8oF:-Kpv1km_FLVWgTIdO8oG:b-s5j48s) ([raw file](https://github.com/robotlocomotion/drake/blob/ab8c0949b0aef74a55f9acaf6006796be443aaaa/drake/systems/primitives/constant_vector_source.cc#L33)):*
> ```cpp
> void ConstantVectorSource<T>::DoCalcVectorOutput(
>     const Context<T>& context, Eigen::VectorBlock<VectorX<T>>* output) const {
>   *output = this->GetNumericParameter(context, source_value_index_).get_value();
> ```

FYI Consider using `*output = get_source_value(context).get_value();` to reduce redundancy.

---

*[drake/systems/primitives/constant_vector_source.h, line 22 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6660#-Kpv01LOoPJkt6uNkksu:-Kpv01LOoPJkt6uNkksv:b-i2apvm) ([raw file](https://github.com/robotlocomotion/drake/blob/ab8c0949b0aef74a55f9acaf6006796be443aaaa/drake/systems/primitives/constant_vector_source.h#L22)):*
> ```objc
> ///
> /// They are already available to link against in the containing library.
> /// No other values for T are currently supported.
> ```

Probably something in the class overview, constructor(s), or both should explain that the value is a declared parameter exposed through the framework.

---

*[drake/systems/primitives/constant_vector_source.h, line 61 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6660#-Kpv1Je8w_gW6ezCi4BT:-Kpv1Je8w_gW6ezCi4BU:b-mypupo) ([raw file](https://github.com/robotlocomotion/drake/blob/ab8c0949b0aef74a55f9acaf6006796be443aaaa/drake/systems/primitives/constant_vector_source.h#L61)):*
> ```objc
>       Eigen::VectorBlock<VectorX<T>>* output) const override;
> 
>   int source_value_index_;
> ```

Missing `{}` on non-const POD member field.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6660#-:-Kpv-gg2Fy6X1lSGgMTG:b-ck7r6o)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/systems/primitives/constant_vector_source.h, line 53 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6660#-Kpv5Mf5tPDfBgcMA2i9:-Kpv5Mf6lWaX5U0Y0Wml:b-9c04j0) ([raw file](https://github.com/robotlocomotion/drake/blob/ab8c0949b0aef74a55f9acaf6006796be443aaaa/drake/systems/primitives/constant_vector_source.h#L53)):*
> ```objc
>   /// Return a mutable pointer to the source value of this block in the given
>   /// @p context
>   BasicVector<T>* get_mutable_source_value(Context<T>* context);
> ```

BTW We seem to have inconsistency on `get_mutable_...()` in that some return references, while others return pointers.

@jwnimmer-tri Do you have any preference between the two? Is it worth anything to bring this up in a separate issue?

---

*[drake/systems/primitives/test/constant_vector_source_test.cc, line 56 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6660#-Kpv5uINR-2Y4E_F7rIg:-Kpv5uIO2KbL0kLMqzb3:bc7ki9z) ([raw file](https://github.com/robotlocomotion/drake/blob/ab8c0949b0aef74a55f9acaf6006796be443aaaa/drake/systems/primitives/test/constant_vector_source_test.cc#L56)):*
> ```cpp
> 
>   // Tests that the output reflects changes to the parameter value in the
>   // context
> ```

nit Missing period

---

*[drake/systems/primitives/test/constant_vector_source_test.cc, line 58 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6660#-Kpv6AgwM6SRzM5tLCwM:-Kpv6AgxoaZX8nRcI-Mg:b-69tlxl) ([raw file](https://github.com/robotlocomotion/drake/blob/ab8c0949b0aef74a55f9acaf6006796be443aaaa/drake/systems/primitives/test/constant_vector_source_test.cc#L58)):*
> ```cpp
>   // context
>   BasicVector<double>* source_value =
>       static_cast<ConstantVectorSource<double>*>(source_.get())
> ```

BTW May or may not be useful to use `dynamic_cast`. GSG says to try and avoid it, we don't have exceptions for it, but we use it quite a bit throughout both core code and in test code.

(May be consistent with the usage down below, `dynamic_cast<const MyVector<...>*>`)

---

*[drake/systems/primitives/test/constant_vector_source_test.cc, line 84 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6660#-Kpv5wy4EEAzmJvU5kts:-Kpv5wy4EEAzmJvU5ktt:bc7ki9z) ([raw file](https://github.com/robotlocomotion/drake/blob/ab8c0949b0aef74a55f9acaf6006796be443aaaa/drake/systems/primitives/test/constant_vector_source_test.cc#L84)):*
> ```cpp
> 
>   // Tests that the output reflects changes to the parameter value in the
>   // context
> ```

nit Missing period

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6660#-:-Kpv7I-4w57U4KxvlO2d:bnfp4nl)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/systems/primitives/constant_vector_source.h, line 48 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6660#-Kpv7Zz9jFoq3zs8dNhZ:-Kpv7ZzA1jHo3qyGEbpC:b-tijiwt) ([raw file](https://github.com/robotlocomotion/drake/blob/ab8c0949b0aef74a55f9acaf6006796be443aaaa/drake/systems/primitives/constant_vector_source.h#L48)):*
> ```objc
> 
>   /// Return a read-only reference to the source value of this block in the
>   /// given @p context
> ```

nit Missing period here and immediately following this as well.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6660)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
+@sammy-tri for follow-up platform review, please.

---

Review status: all files reviewed at latest revision, 9 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6660#-:-Kpv7rU-Swfrl1uZ1POi:b-9eeq1w)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/systems/primitives/constant_vector_source.h, line 53 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6660#-Kpv5Mf5tPDfBgcMA2i9:-Kpv8LNksPmbkbvG_KuW:b-n099r3) ([raw file](https://github.com/robotlocomotion/drake/blob/ab8c0949b0aef74a55f9acaf6006796be443aaaa/drake/systems/primitives/constant_vector_source.h#L53)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW We seem to have inconsistency on `get_mutable_...()` in that some return references, while others return pointers.

@jwnimmer-tri Do you have any preference between the two? Is it worth anything to bring this up in a separate issue?
</blockquote></details>

This is a good question.  I know it has been discussed, but I don't remember the consensus off the top of my head.  In this case, the API is mimicing the System method (just binding in the relevant `int`), so my vote would be for consistency (leave it as-is).  If or when System changes its API, this method should follow suit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6660)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/primitives/constant_vector_source.h, line 53 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6660#-Kpv5Mf5tPDfBgcMA2i9:-Kpv9-K6ePoCmoKcivSj:b-81nlyc) ([raw file](https://github.com/robotlocomotion/drake/blob/ab8c0949b0aef74a55f9acaf6006796be443aaaa/drake/systems/primitives/constant_vector_source.h#L53)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This is a good question.  I know it has been discussed, but I don't remember the consensus off the top of my head.  In this case, the API is mimicing the System method (just binding in the relevant `int`), so my vote would be for consistency (leave it as-is).  If or when System changes its API, this method should follow suit.
</blockquote></details>

SGTM.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6660)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> pending resolution of the existing items

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6660#-:-KpvZ-HGMpc2B84JQemk:bdcimkn)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/primitives/constant_vector_source.h, line 53 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6660#-Kpv5Mf5tPDfBgcMA2i9:-Kpv_BfD3FgXe-Uwu366:b-z9pwvr) ([raw file](https://github.com/robotlocomotion/drake/blob/ab8c0949b0aef74a55f9acaf6006796be443aaaa/drake/systems/primitives/constant_vector_source.h#L53)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

SGTM.
</blockquote></details>

FYI, my preference is mutable reference returns whenever possible (which is whenever nullptr cannot be returned). I don't know if there is a controversy about that but we have definitely been inconsistent. My reasons:
- pointers should be checked for null; references don't require that.
- Google return value style doesn't work correctly with mutable pointer returns. Consider `f(i,j,&get_mutable_blah())` vs. `f(i,j,get_mutable-blah())`. The `&` syntax to indicate output parameters is the justification for Google's "mutable arguments are pointers" rule; seems a shame to defeat it.
- general disgust for naked pointers.

I would vote for not adding any more pointer returns rather than being consistent with System. That would help tip the scale towards the reference returns.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6660)*
<!-- Sent from Reviewable.io -->

avalenzu commented:
Review status: 0 of 3 files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/systems/primitives/constant_vector_source.cc, line 33 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6660#-Kpv1km_FLVWgTIdO8oF:-Kq-Wvb5cYexB7tZmJCM:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ab8c0949b0aef74a55f9acaf6006796be443aaaa/drake/systems/primitives/constant_vector_source.cc#L33)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI Consider using `*output = get_source_value(context).get_value();` to reduce redundancy.
</blockquote></details>

Done.

---

*[drake/systems/primitives/constant_vector_source.h, line 22 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6660#-Kpv01LOoPJkt6uNkksu:-Kq3gohY_L_djzPCvgxb:b-qsvnoa) ([raw file](https://github.com/robotlocomotion/drake/blob/ab8c0949b0aef74a55f9acaf6006796be443aaaa/drake/systems/primitives/constant_vector_source.h#L22)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Probably something in the class overview, constructor(s), or both should explain that the value is a declared parameter exposed through the framework.
</blockquote></details>

Done. PTAL.

---

*[drake/systems/primitives/constant_vector_source.h, line 48 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6660#-Kpv7Zz9jFoq3zs8dNhZ:-Kq3V4db0s3qzMujPoh_:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ab8c0949b0aef74a55f9acaf6006796be443aaaa/drake/systems/primitives/constant_vector_source.h#L48)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

nit Missing period here and immediately following this as well.
</blockquote></details>

Done.

---

*[drake/systems/primitives/constant_vector_source.h, line 53 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6660#-Kpv5Mf5tPDfBgcMA2i9:-Kq3gw7vzaF5fUIHU80j:byn3an1) ([raw file](https://github.com/robotlocomotion/drake/blob/ab8c0949b0aef74a55f9acaf6006796be443aaaa/drake/systems/primitives/constant_vector_source.h#L53)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

FYI, my preference is mutable reference returns whenever possible (which is whenever nullptr cannot be returned). I don't know if there is a controversy about that but we have definitely been inconsistent. My reasons:
- pointers should be checked for null; references don't require that.
- Google return value style doesn't work correctly with mutable pointer returns. Consider `f(i,j,&get_mutable_blah())` vs. `f(i,j,get_mutable-blah())`. The `&` syntax to indicate output parameters is the justification for Google's "mutable arguments are pointers" rule; seems a shame to defeat it.
- general disgust for naked pointers.

I would vote for not adding any more pointer returns rather than being consistent with System. That would help tip the scale towards the reference returns.
</blockquote></details>

I'm going to leave this consistent with the System method. If we want to change the convention used in the System framework, I'd be all for it, but I think it should happen as its own change.

---

*[drake/systems/primitives/constant_vector_source.h, line 61 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6660#-Kpv1Je8w_gW6ezCi4BT:-Kq3V5CPoc7pWdM120Hp:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ab8c0949b0aef74a55f9acaf6006796be443aaaa/drake/systems/primitives/constant_vector_source.h#L61)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Missing `{}` on non-const POD member field.
</blockquote></details>

Done.

---

*[drake/systems/primitives/test/constant_vector_source_test.cc, line 56 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6660#-Kpv5uINR-2Y4E_F7rIg:-Kq3VjhVhbGKLennwz2y:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ab8c0949b0aef74a55f9acaf6006796be443aaaa/drake/systems/primitives/test/constant_vector_source_test.cc#L56)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

nit Missing period
</blockquote></details>

Done.

---

*[drake/systems/primitives/test/constant_vector_source_test.cc, line 58 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6660#-Kpv6AgwM6SRzM5tLCwM:-Kq3Vaj08HO1O7_ZLo0B:bic0j0v) ([raw file](https://github.com/robotlocomotion/drake/blob/ab8c0949b0aef74a55f9acaf6006796be443aaaa/drake/systems/primitives/test/constant_vector_source_test.cc#L58)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW May or may not be useful to use `dynamic_cast`. GSG says to try and avoid it, we don't have exceptions for it, but we use it quite a bit throughout both core code and in test code.

(May be consistent with the usage down below, `dynamic_cast<const MyVector<...>*>`)
</blockquote></details>

Sticking with `static_cast`.

---

*[drake/systems/primitives/test/constant_vector_source_test.cc, line 84 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6660#-Kpv5wy4EEAzmJvU5kts:-Kq3Vzd8P_FCisO1H7Mm:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ab8c0949b0aef74a55f9acaf6006796be443aaaa/drake/systems/primitives/test/constant_vector_source_test.cc#L84)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

nit Missing period
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6660)*
<!-- Sent from Reviewable.io -->

avalenzu commented:
Addressed Reviewer comments. @jwnimmer-tri , please let me know if the updated class description is sufficient.

---

Review status: 0 of 3 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6660#-:-Kq3huNNzcjtoOKhGvFK:b-g4dhmj)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6660)*
<!-- Sent from Reviewable.io -->
